Port the frame output example to use the Document API

nc4rrillo opened the pull request:
Takes care of #1695 by replacing the iframe with an external image which gets the output of a sub document piped to it.

I had to change the ordering out the `set_output_image_handler` and `set_external_image_handler` so that I'd have a generated texture ID to pass to the `ExternalImageHandler` struct.

This removes all GL calls besides generating the texture id. 

r? @glennw @kvark

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/2073)
<!-- Reviewable:end -->

kvark commented:
Great work, with somewhat unexpected timing (we just merged the document API) :)

> I had to change the ordering out the set_output_image_handler and set_external_image_handler so that I'd have a generated texture ID to pass to the ExternalImageHandler struct.

Uh oh, that's pretty hacky.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[webrender/examples/frame_output.rs, line 37 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUkFTNQCr2Brw9GEip:-KzUkFTOvpYdAwvFBmC-:b-vborwo) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L37)):*
> ```Rust
> 
> struct OutputHandler {
>     texture_id: gl::GLuint
> ```

we should really re-export this type (and/or wrap it) in order to avoid exposing GL's internals to the outside world (too much)

---

*[webrender/examples/frame_output.rs, line 52 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUkaEybx7Fk0WruylB:-KzUkaEzLmEsM0xc_MR9:b-i31rcl) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L52)):*
> ```Rust
> impl ExternalHandler {
>     fn new(texture_id: gl::GLuint) -> ExternalHandler {
>         ExternalHandler { texture_id }
> ```

these constructors are not too useful, to be honest

---

*[webrender/examples/frame_output.rs, line 99 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUl4dMJEkusviQWjjA:-KzUl4dMJEkusviQWjjB:b-5ls5cy) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L99)):*
> ```Rust
> 
>         let pipeline_id = PipelineId(1, 0);
>         let layer = -1;
> ```

hmm, I'm confused. This one should be on top, and the rendered documents are sorted in ascending order for rendering. Given that the boilerplate-created one has the layer of 0, it will go on top of this, will it not?

---

*[webrender/examples/frame_output.rs, line 103 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUj-wbfW2tB-nibxeL:-KzUj-wcJahpliUb7ERP:b1c2y87) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L103)):*
> ```Rust
>         let bounds = DeviceUintRect::new(DeviceUintPoint::zero(), framebuffer_size);
>         let document_id = api.add_document(framebuffer_size, layer);
>         
> ```

nit: spaces

---

*[webrender/examples/frame_output.rs, line 142 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUj9C1rknHmd24E5Ro:-KzUj9C2SPE2zfJMsLFe:b-kjrh5a) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L142)):*
> ```Rust
>             resources,
>         );
>         
> ```

spaces

---

*[webrender/examples/frame_output.rs, line 159 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUjQhRVMi7ivafIORY:-KzUjQhRVMi7ivafIORZ:bizawzt) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L159)):*
> ```Rust
>     ) {
>         if self.output_document.is_none(){
>             let device_pixel_ratio = framebuffer_size.width as f32 /
> ```

that strange feeling... when your hack starts multiplying itself :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2073#-:-KzUgg4RBbu1tCyOjYCi:b-tuzqkd)*
<!-- Sent from Reviewable.io -->

nc4rrillo commented:
Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[webrender/examples/frame_output.rs, line 37 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUkFTNQCr2Brw9GEip:-KzUt63ayKRZQ5ozyW3l:bt8vc3r) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L37)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

we should really re-export this type (and/or wrap it) in order to avoid exposing GL's internals to the outside world (too much)
</blockquote></details>

Is there any reason the caller (in this case the Example App) needs to provide the texture id? Can't WR assign one, abstracting away the GL usage? 

---

*[webrender/examples/frame_output.rs, line 52 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUkaEybx7Fk0WruylB:-KzUsqddMjBEyyB5P1fc:b-886gsx) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L52)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

these constructors are not too useful, to be honest
</blockquote></details>

Fixed

---

*[webrender/examples/frame_output.rs, line 99 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUl4dMJEkusviQWjjA:-KzUswiaLogpGIwHJUpo:by7q3d) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L99)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

hmm, I'm confused. This one should be on top, and the rendered documents are sorted in ascending order for rendering. Given that the boilerplate-created one has the layer of 0, it will go on top of this, will it not?
</blockquote></details>

Fixed. This was a hold over when I was still experimenting with the document API

---

*[webrender/examples/frame_output.rs, line 103 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUj-wbfW2tB-nibxeL:-KzUt-s_SyOgpPhU41M5:b-886gsx) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L103)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: spaces
</blockquote></details>

Fixed

---

*[webrender/examples/frame_output.rs, line 142 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUj9C1rknHmd24E5Ro:-KzUt3OtZ-oUpYW7u4UH:b-886gsx) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L142)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

spaces
</blockquote></details>

Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2073)*
<!-- Sent from Reviewable.io -->

nc4rrillo commented:
Yeah I don't like this much. Any ideas for a less hacky solution? 

---

Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2073#-:-KzUuyrEXosEWHqLC6Am:b91q13o)*
<!-- Sent from Reviewable.io -->

kvark commented:
(proposed in the comment)

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[webrender/examples/frame_output.rs, line 37 at r1](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUkFTNQCr2Brw9GEip:-KzUwKkVwzUiK8xnxoUV:b-vzgs3y) ([raw file](https://github.com/servo/webrender/blob/f2e1ff45c629321032bebc5d2ea2e73eaa7a1d5c/webrender/examples/frame_output.rs#L37)):*
<details><summary><i>Previously, nc4rrillo (Nelson Carrillo) wrote…</i></summary><blockquote>

Is there any reason the caller (in this case the Example App) needs to provide the texture id? Can't WR assign one, abstracting away the GL usage? 
</blockquote></details>

In the long(er) term, we should rethink/improve the frame output API. In the short term, we should provide a simple helper function to give us that GL texture in order to abstract it.

---

*[webrender/examples/common/boilerplate.rs, line 155 at r2](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUws981DQ7LzIqV_HY:-KzUws981DQ7LzIqV_HZ:b52zbi7) ([raw file](https://github.com/servo/webrender/blob/9ad748a7cc17325b1d3895e2addb88dcc1acad51/webrender/examples/common/boilerplate.rs#L155)):*
> ```Rust
>     }
> 
>     if let Some(external_image_handler) = example.get_external_image_handler() {
> ```

let's just have `get_image_handlers` returning a struct/tuple, moving the ordering requirements to the example itself

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2073#-:-KzUx-C3RGOJw7zd0yA8:be3rycd)*
<!-- Sent from Reviewable.io -->

nc4rrillo commented:
Done.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[webrender/examples/common/boilerplate.rs, line 155 at r2](https://reviewable.io:443/reviews/servo/webrender/2073#-KzUws981DQ7LzIqV_HY:-KzV4MjMUj5pTYdOqF-_:bb6un1j) ([raw file](https://github.com/servo/webrender/blob/9ad748a7cc17325b1d3895e2addb88dcc1acad51/webrender/examples/common/boilerplate.rs#L155)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

let's just have `get_image_handlers` returning a struct/tuple, moving the ordering requirements to the example itself
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2073#-:-KzV4IrIw4K_I-FdlhmC:b-896fix)*
<!-- Sent from Reviewable.io -->

kvark commented:
thank you! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2073#-:-KzVNxMUeacUGf-IsFjA:bfl09gv)*
<!-- Sent from Reviewable.io -->

kvark commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit dc1f896 has been approved by `kvark`

<!-- @bors-servo r=kvark dc1f896398a6e3002877691f8d680a06153cdff8 -->

bors-servo commented:
:hourglass: Testing commit dc1f896398a6e3002877691f8d680a06153cdff8 with merge 24088e025e26e76bf3bc1ce5ccf0596b7c32716e...

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/webrender/build/1.0.6561), [status-travis](https://travis-ci.org/servo/webrender/builds/305500940?utm_source=github_status&utm_medium=notification)
Approved by: kvark
Pushing 24088e025e26e76bf3bc1ce5ccf0596b7c32716e to master...
